craft_application/git/_git_repo.py

```diff
@@ -308,3 +405,46 @@ def _resolve_ref(self, ref: str) -> str:
             ) from error
         logger.debug("Resolved reference %r for name %r", reference, ref)
         return reference
+
+    @classmethod
+    def clone_repository(
+        cls,
+        *,
+        url: str,
+        path: Path,
+        checkout_branch: str | None = None,
+        depth: int = 0,
+    ) -> GitRepo:
+        """Clone repository to specific path and return GitRepo object.
+
+        :param url: the URL to clone repository from
+        :param path: path to the directory where repository should be cloned
+        :param checkout_branch: optional branch which should be checked out
+
+        :returns: GitRepo wrapper around the repository
+
+        raises GitError: if cloning repository cannot be done
+        """
+        if is_repo(path):
+            raise GitError("Cannot clone to existing repository")
+
+        logger.debug("Cloning %s to %s", url, path)
+        if checkout_branch is not None:
+            logger.debug("Checking out to branch: %s", checkout_branch)
+
+        try:
+            # TODO: consider using git CLI with --single-branch
+            #  as it is not implemented in pygit2
+            #  https://github.com/libgit2/pygit2/issues/1290
+            pygit2.clone_repository(
+                url,
+                path=os.fspath(path),
+                checkout_branch=checkout_branch,
+                depth=depth,
+                callbacks=GitLogCallbacks(),
+            )
+        except KeyError as ke:
+            raise GitError(f"cannot find branch `{checkout_branch}` in {url}") from ke
+        except pygit2.GitError as ge:
+            raise GitError(f"cannot clone repository: {url} to {str(path)!r}") from ge
+        return cls(path)
```

Comment:
Rewrite to subprocess, due to performance implications

Reply:
Ended up using `process_run` from craft_parts.utils.os_utils and checking if the function is called with correct arguments in tests